File: bright_vision_core/spec_layers.py
"""Heuristics and normalization for three-layer generated specs."""

from __future__ import annotations

import re


def design_references_requirements(requirements: str, design: str) -> bool:
    req = (requirements or "").strip()
    des = (design or "").strip()
    if not des or not re.search(r"REQ-\d+", req, re.I):
        return True
    if re.search(r"REQ-\d+", des, re.I):
        return True
    nums = [m.group(1) for m in re.finditer(r"REQ-(\d+)", req, re.I)]
    if any(re.search(rf"\b{n}\b", des) for n in nums):
        return True
    if re.search(r"\brequirement\s*\d+", des, re.I):
        return True
    return False


def requirement_ids(requirements: str) -> list[str]:
    return list(dict.fromkeys(re.findall(r"REQ-\d+", requirements, re.I)))


def normalize_spec_layer_traceability(layers: dict[str, str]) -> dict[str, str]:
    """Ensure design cites REQ ids when requirements define them (small-model guard)."""
    req = (layers.get("requirements") or "").strip()
    design = (layers.get("design") or "").strip()
    ids = requirement_ids(req)
    if not ids or design_references_requirements(req, design):
        return layers
    trace = "Covers " + ", ".join(ids) + "."
    out = dict(layers)
    if not design:
        out["design"] = f"## Traceability\n{trace}"
    else:
        out["design"] = f"{design.rstrip()}\n\n## Traceability\n{trace}"
    return out


def assess_generated_spec_layers(
    requirements: str,
    design: str,
    tasks_md: str,
) -> tuple[bool, list[str]]:
    issues: list[str] = []
    req = (requirements or "").strip()
    des = (design or "").strip()
    tasks = (tasks_md or "").strip()

    if not req:
        issues.append("requirements empty")
    if not des:
        issues.append("design empty")
    if not tasks:
        issues.append("tasks_md empty")

    if req:
        if not re.search(r"REQ-\d+", req, re.I):
            issues.append("requirements missing REQ-### id")
        if not re.search(r"\bshall\b", req, re.I):
            issues.append("requirements missing SHALL")
        if not re.search(r"\bwhen\b", req, re.I):
            issues.append("requirements missing WHEN")

    if tasks and not re.search(r"(?:^|\n)\s*(?:-\s*\[[ xX]\]\s*)?\d+\.\s+", tasks):
        issues.append("tasks_md missing numbered implementation steps")

    if des and req and not design_references_requirements(req, des):
        if not (tasks and design_references_requirements(req, tasks)):
            issues.append("design does not reference any REQ id")

    return len(issues) == 0, issues

File: src/utils/specLayers.ts
/**
 * Heuristic checks that generated three-layer specs are usable (EARS + tasks).
 * Shared by unit tests and Playwright e2e.
 */

export interface SpecLayerAssessment {
  ok: boolean
  issues: string[]
}

export interface SpecLayers {
  requirements: string
  design: string
  tasks_md: string
}

/** True when design text traces back to REQ ids (or equivalent) in requirements. */
export function designReferencesRequirements(requirements: string, design: string): boolean {
  const req = (requirements || '').trim()
  const des = (design || '').trim()
  if (!des || !/REQ-\d+/i.test(req)) return true
  if (/REQ-\d+/i.test(des)) return true
  const nums = [...req.matchAll(/REQ-(\d+)/gi)].map((m) => m[1]).filter(Boolean)
  if (nums.some((n) => new RegExp(`\\b${n}\\b`).test(des))) return true
  if (/\brequirement\s*\d+/i.test(des)) return true
  return false
}

export function assessGeneratedSpecLayers(layers: SpecLayers): SpecLayerAssessment {
  const issues: string[] = []
  const req = (layers.requirements || '').trim()
  const design = (layers.design || '').trim()
  const tasks = (layers.tasks_md || '').trim()

  if (!req) issues.push('requirements empty')
  if (!design) issues.push('design empty')
  if (!tasks) issues.push('tasks_md empty')

  if (req) {
    if (!/REQ-\d+/i.test(req)) issues.push('requirements missing REQ-### id')
    if (!/\bshall\b/i.test(req)) issues.push('requirements missing SHALL')
    if (!/\bwhen\b/i.test(req)) issues.push('requirements missing WHEN')
  }

  if (tasks && !/(?:^\s*[-*]\s*\[[ xX]\]\s*)?\d+\.\s+/m.test(tasks)) {
    issues.push('tasks_md missing numbered implementation steps')
  }

  if (design && req && !designReferencesRequirements(req, design)) {
    if (!(tasks && designReferencesRequirements(req, tasks))) {
      issues.push('design does not reference any REQ id')
    }
  }

  return { ok: issues.length === 0, issues }
}

/** Mock / fixture content that passes {@link assessGeneratedSpecLayers}. */
export const MOCK_SANE_SPEC_LAYERS: SpecLayers = {
  requirements: `### REQ-001
**WHEN** the user opens the feature
**THE** system **SHALL** show the task spec layers.

### REQ-002
**WHEN** the user saves requirements
**THE** system **SHALL** persist markdown under \`.cecli/specs/{id}/\`.`,
  design: `## Overview
Covers REQ-001 UI flow and REQ-002 disk sync.`,
  tasks_md: `- [ ] 1. Wire generate-spec API for REQ-001 (depends: none)
- [ ] 2. Add e2e coverage for REQ-002 (depends: 1)`,
}
